pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java

```diff
@@ -119,52 +118,56 @@ public ConsumerBuilderImpl(PulsarClientImpl client, Schema<T> schema) {
-            try {
-                if (client.getPartitionedTopicMetadata(oldRetryLetterTopic)
-                        .get(client.conf.getOperationTimeoutMs(), TimeUnit.MILLISECONDS).partitions > 0) {
-                    retryLetterTopic = oldRetryLetterTopic;
-                }
-                if (client.getPartitionedTopicMetadata(oldDeadLetterTopic)
-                        .get(client.conf.getOperationTimeoutMs(), TimeUnit.MILLISECONDS).partitions > 0) {
-                    deadLetterTopic = oldDeadLetterTopic;
-                }
-            } catch (InterruptedException | TimeoutException e) {
-                return FutureUtil.failedFuture(e);
-            } catch (ExecutionException e) {
-                return FutureUtil.failedFuture(e.getCause());
-            }
-
-            if (conf.getDeadLetterPolicy() == null) {
-                conf.setDeadLetterPolicy(DeadLetterPolicy.builder()
-                                        .maxRedeliverCount(RetryMessageUtil.MAX_RECONSUMETIMES)
-                                        .retryLetterTopic(retryLetterTopic)
-                                        .deadLetterTopic(deadLetterTopic)
-                                        .build());
+            CompletableFuture<PartitionedTopicMetadata> retryLetterTopicMetadata =
+                    client.getPartitionedTopicMetadata(oldRetryLetterTopic);
+            CompletableFuture<PartitionedTopicMetadata> deadLetterTopicMetadata =
+                    client.getPartitionedTopicMetadata(oldDeadLetterTopic);
+            applyDLQConfig = CompletableFuture.allOf(retryLetterTopicMetadata, deadLetterTopicMetadata)
+                    .thenAccept(__ -> {
+                        String retryLetterTopic = topicFirst + "-" + conf.getSubscriptionName()
+                                        + RetryMessageUtil.RETRY_GROUP_TOPIC_SUFFIX;
+                        String deadLetterTopic = topicFirst + "-" + conf.getSubscriptionName()
+                                        + RetryMessageUtil.DLQ_GROUP_TOPIC_SUFFIX;
+                        if (retryLetterTopicMetadata.join().partitions > 0) {
+                            retryLetterTopic = oldRetryLetterTopic;
+                        }
+                        if (deadLetterTopicMetadata.join().partitions > 0) {
+                            deadLetterTopic = oldDeadLetterTopic;
+                        }
+                        if (conf.getDeadLetterPolicy() == null) {
+                            conf.setDeadLetterPolicy(DeadLetterPolicy.builder()
+                                    .maxRedeliverCount(RetryMessageUtil.MAX_RECONSUMETIMES)
+                                    .retryLetterTopic(retryLetterTopic)
+                                    .deadLetterTopic(deadLetterTopic)
+                                    .build());
+                        } else {
+                            if (StringUtils.isBlank(conf.getDeadLetterPolicy().getRetryLetterTopic())) {
```

Comment:
> if (StringUtils.isBlank(conf.getDeadLetterPolicy().getRetryLetterTopic()))

Is it worth doing this check earlier before doing the lookup? If the caller specifies these directly, then we're paying the cost of two metadata calls that are then discarded.

Reply:
Yes, right. Thanks for reviewing.